runtime/interpreter/decode.go

```diff
@@ -1736,3 +1625,186 @@ func (d *DecoderV4) decodeCapabilityStaticType() (StaticType, error) {
+// Meta info includes:
+//    - location
+//    - qualifiedIdentifier
+//    - kind
+//
+// This also extracts out the fields raw content and cache it separately inside the value.
+//
+func decodeCompositeMetaInfo(v *CompositeValue, content []byte) error {
+	d, err := NewByteDecoder(content, v.Owner, CurrentEncodingVersion, v.decodeCallback)
+	if err != nil {
+		return err
+	}
+
+	const expectedLength = encodedCompositeValueLength
+
+	size, err := d.decoder.DecodeArrayHead()
+
+	if err != nil {
+		if e, ok := err.(*cbor.WrongTypeError); ok {
+			return fmt.Errorf("invalid composite encoding (@ %s): expected [%d]interface{}, got %s",
+				strings.Join(v.valuePath, "."),
+				expectedLength,
+				e.ActualType.String(),
+			)
+		}
+		return err
+	}
+
+	if size != expectedLength {
+		return fmt.Errorf("invalid composite encoding (@ %s): expected [%d]interface{}, got [%d]interface{}",
+			strings.Join(v.valuePath, "."),
+			expectedLength,
+			size,
+		)
+	}
+
+	// Location
+
+	// Decode location at array index encodedCompositeValueLocationFieldKey
+	location, err := d.decodeLocation()
+	if err != nil {
+		return fmt.Errorf(
+			"invalid composite location encoding (@ %s): %w",
+			strings.Join(v.valuePath, "."),
+			err,
+		)
+	}
+
+	// Skip obsolete element at array index encodedCompositeValueTypeIDFieldKey
+	err = d.decoder.Skip()
+	if err != nil {
+		return err
+	}
+
+	// Kind
+
+	// Decode kind at array index encodedCompositeValueKindFieldKey
+	encodedKind, err := d.decoder.DecodeUint64()
+	if err != nil {
+		if e, ok := err.(*cbor.WrongTypeError); ok {
+			return fmt.Errorf(
+				"invalid composite kind encoding (@ %s): %s",
+				strings.Join(v.valuePath, "."),
+				e.ActualType.String(),
+			)
+		}
+		return err
+	}
+
+	kind := common.CompositeKind(encodedKind)
+
+	// Fields
+
+	fieldsContent, err := d.decoder.DecodeRawBytes()
+	if err != nil {
+		if e, ok := err.(*cbor.WrongTypeError); ok {
+			return fmt.Errorf(
+				"invalid composite fields encoding (@ %s): %s",
+				strings.Join(v.valuePath, "."),
+				e.ActualType.String(),
+			)
+		}
+		return err
+	}
+
+	// Qualified identifier
+
+	// Decode qualified identifier at array index encodedCompositeValueQualifiedIdentifierFieldKey
+	qualifiedIdentifier, err := d.decoder.DecodeString()
+	if err != nil {
+		if e, ok := err.(*cbor.WrongTypeError); ok {
+			return fmt.Errorf(
+				"invalid composite qualified identifier encoding (@ %s): %s",
+				strings.Join(v.valuePath, "."),
+				e.ActualType.String(),
+			)
+		}
+		return err
+	}
+
+	v.location = location
+	v.qualifiedIdentifier = qualifiedIdentifier
+	v.kind = kind
+	v.fieldsContent = fieldsContent
+
+	return nil
+}
+
+// decodeCompositeFields decodes fields from the byte content and updates the composite value.
+//
+func decodeCompositeFields(v *CompositeValue, content []byte) error {
+	d, err := NewByteDecoder(content, v.Owner, CurrentEncodingVersion, v.decodeCallback)
```

Comment:
The version of the data might not be the current version. The version of the encoded data should be saved in the deferred composite and then used here